TransformManager.cs

```diff
@@ -348,87 +320,141 @@ protected void AddGizmo( List<SceneObject> targets )
-            if (l1.Count != l2.Count)
-                return false;
-            foreach(T o in l1) {
-                if (!l2.Contains(o))
-                    return false;
+            if (targets.Count == 1 && _objectFilters.Count > 0)
+            {
+                foreach (GizmoTypeFilter filter in _objectFilters)
+                {
+                    string typeName = filter.FilterF(targets[0]);
+                    if (typeName != null && _gizmoTypes.ContainsKey(typeName))
+                    {
+                        return _gizmoTypes[typeName];
+                    }
+                }
             }
-            return true;
+
+            return defaultBuilder;
         }
 
+        private static bool UnorderedListsEqual<T>(IReadOnlyCollection<T> l1, ICollection<T> l2) => 
+            l1.Count == l2.Count && l1.All(l2.Contains);
 
         private void Scene_SelectionChangedEvent(object sender, EventArgs e)
         {
             FScene scene = Context.Scene;
             List<SceneObject> vSelected = new List<SceneObject>();
-            foreach ( SceneObject tso in scene.Selected ) {
-                if (tso != null) {
-                    if ( SelectionFilterF == null || SelectionFilterF(tso) )
-                        vSelected.Add(tso);
+            foreach (SceneObject tso in scene.Selected)
+            {
+                if (tso != null && (_selectionFilterF == null || _selectionFilterF(tso)))
+                {
+                    vSelected.Add(tso);
                 }
             }
 
-            if (vSelected.Count == 0 && activeGizmo != null) {
+            if (vSelected.Count == 0 && _activeGizmo != null)
+            {
                 // object de-selected, dismiss gizmo
                 DismissActiveGizmo();
                 return;
             }
-            if (activeGizmo != null && unordered_lists_equal(vSelected, activeGizmo.Targets) == false) {
+
+            if (_activeGizmo != null && UnorderedListsEqual(vSelected, _activeGizmo.Targets) == false)
+            {
                 DismissActiveGizmo();
             }
 
             if (vSelected.Count > 0)
-                Context.RegisterNextFrameAction(add_gizmo_next_frame);
-                //AddGizmo(vSelected);
+            {
+                Context.RegisterNextFrameAction(AddGizmoNextFrame);
+            }
+            //AddGizmo(vSelected);
         }
 
-
-
-        private void add_gizmo_next_frame()
+        private void AddGizmoNextFrame()
         {
             FScene scene = Context.Scene;
             List<SceneObject> vSelected = new List<SceneObject>();
-            foreach (SceneObject tso in scene.Selected) {
-                if (tso != null) {
-                    if (SelectionFilterF == null || SelectionFilterF(tso))
-                        vSelected.Add(tso);
+            foreach (SceneObject tso in scene.Selected)
+            {
+                if (tso != null && (_selectionFilterF == null || _selectionFilterF(tso)))
```

Comment:
The same logic in 391-393. Maybe it is better to move this logic to new method and use it.

Reply:
done